### ISSUE TITLE
Asdf - yarn via corepack compatibility

### ISSUE THREAD

**Issue by Maxim-Filimonov** [asdf-vm/asdf-nodejs#342]
I'm wondering with upcoming corepack feature stabilization are there any ways to integrate it with asdf?

Right now, it seem when we do `corepack enable` in one repo it enables it for all the repos which share the same version of node. 
We have decided to opt out of corepack for now to avoid upgrading yarn across repos. 

What I envision integration would allow is to separate yarn version for each repo like we do with asdf now but able to enable corepact without affecting other repos.

**Comment by augustobmoura** [comment]
I still need to understand how corepack works behind the scenes, but we are probably willing to offer a better integration. Contributions are also welcome

**Comment by Techn1x** [comment]
> Right now, it seem when we do corepack enable in one repo it enables it for all the repos which share the same version of node.

I'm using pnpm though but otherwise I'm seeing the same thing.

Since corepack just dumps the shims in the same dir as `node` https://github.com/nodejs/corepack#corepack-enable--name I worry that asdf plugins like `yarn` and `pnpm` would cause issues here too (though I don't know for sure)

Like for a different repo, if I have `nodejs` and `pnpm` plugins, which would also create a `pnpm` shim in the shims folder - would that interfere with some other repo that gets pnpm via corepack, or vice versa

**Comment by jdxcode** [comment]
I believe you would need to put asdf-yarn/asdf-pnpm would need to be specified before node in order for the corepack shims from asdf-nodejs to not be used

**Comment by Maxim-Filimonov** [comment]
> I believe you that asdf-yarn/asdf-pnpm would need to be specified before node in .tool-versions in order for the corepack shims from asdf-nodejs to not be used

we did exactly that for `asdf-yarn` and it worked. However, if any dev enables corepack on other repo it will break this one i believe. Or will it not?

**Comment by akkadaya** [comment]
Does this help?
- Yarn 1 https://classic.yarnpkg.com/lang/en/docs/cli/version
- Yarn >2 https://yarnpkg.com/cli/version

It will let you install Yarn in the project or lock the version

**Comment by dtinth** [comment]
Thanks to this:

- https://github.com/asdf-vm/asdf-nodejs/pull/349/files

The solution I took was:

```sh
corepack enable
corepack reshim nodejs
```

So now, I use `corepack` to manage `yarn`/`pnpm` version and do not use asdf plugins for that.

**Comment by Techn1x** [comment]
Unfortunately some repos in my organisation will likely continue using yarn/pnpm plugins, so I don't think that will work for us.

**Comment by mnahkies** [comment]
@dtinth I think you meant to write `asdf reshim nodejs` but otherwise this solution worked for me

**Comment by dtinth** [comment]
@mnahkies thank you for the correction, edited.

**Comment by stephenh** [comment]
I don't 100% know if this is best practice, but currently I'm using `asdf-nodejs` to manage Node versions across repos via `.nvmrc` (as usual), and then for each repo that needs a new version of node, just doing `corepack enable && asdf reshim nodejs`, to get the right version of `yarn`.

This is working really well, so thanks! Just naively it seems nice to ideally avoid re-running `corepack enable && asdf reshim nodejs` and somehow having `asdf-nodejs` know it should do that for me.

I get there is probably not a good indicator for when this should happen; maybe a new config flag/env variable, or looking for a `packageManager` key in `package.json` (seems neat?), but just mentioning it as an idea. Thanks!

**Comment by katelovescode** [comment]
> Thanks to this:
> 
> * https://github.com/asdf-vm/asdf-nodejs/pull/349/files
> 
> The solution I took was:
> 
> ```shell
> corepack enable
> asdf reshim nodejs
> ```
> 
> So now, I use `corepack` to manage `yarn`/`pnpm` version and do not use asdf plugins for that.

Thanks so much, this is what I needed to do.

**Comment by 1oglop1** [comment]
Sorry for being blunt but I am not very well-versed in what is going on with `corepack`.
It would be great if someone could explain this simply or point me to the right resources.

I've been using ASDF just for `nodejs and pnpm` up until version 18
```
nodejs   18.20.2
pnpm 9.0.1
```
Then this `corepack` change came and ruined my entire setup 🙀 and I had to adopt it unwillingly for one of the projects.

I've been using renovate to keep `.tool-versions` up to date and everything worked great, all environments had tools synchronized. 

After I enabled corepack
```
corepack enable
asdf reshim nodejs
```
it keep using whatever version from `.asdf/installs/nodejs/20.8.0/bin/pnpm` which I found was the version I did not want to use. 

After some digging, I learned that I can set the version in `package.json` -> `"packageManager": "pnpm@9.0.1"`
then after `corepack use pnpm@9.0.1`
Renovate should keep the version up to date according to: https://docs.renovatebot.com/modules/manager/npm/#additional-information

my question is what other do I need to make to ensure the `pnpm` version is what I expect across all projects I work with?

Thank you!

**Comment by dtinth** [comment]
@1oglop1 With Corepack, when you run `pnpm`…

- **Case 1:** In a folder with `package.json` containing `"packageManager": "pnpm@…"`
  - Corepack will use the specified version of `pnpm`.
  - If the specified version has not yet been installed, Corepack will download and install it automatically.
  - To upgrade the version of `pnpm` in this case — that is, in your project — run `corepack use pnpm`. This updates `package.json` file.

- **Case 2:** In a folder with `package.json` containing `"packageManager": "yarn@…"` (something other than `pnpm`)
  - Corepack will refuse to let you run `pnpm` and will say “Usage Error: This project is configured to use yarn” instead.

- **Case 3:** In other cases
  - Corepack will use the globally-installed version of `pnpm`.
  - To upgrade the globally-installed version, run `corepack install --global pnpm`

To summarize

- To keep your globally-installed version updated, run `corepack install --global pnpm`
- To keep the `pnpm` version in your project up-to-date, run `corepack use pnpm`

**Comment by augustobmoura** [comment]
Hello everyone! :wave: 
I just merged a new setting to master that allows corepack to be enabled by default on every installation.

Just set `ASDF_NODEJS_AUTO_ENABLE_COREPACK` on you rc file to anything aside from empty, and it should be good to go!
